fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/reamortization/LoanReAmortizationValidator.java

```diff
@@ -18,18 +18,118 @@
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.supported.only.for.progressive.loan.schedule.type",
+                    "Loan reamortization is only available for progressive repayment schedule and Advanced payment allocation strategy",
+                    loan.getId());
+        }
+
+        // validate reamortization is only available for non-interest bearing loans
+        if (loan.isInterestBearing()) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.supported.only.for.non.interest.loans",
+                    "Loan reamortization is only available for non-interest bearing loans", loan.getId());
+        }
+
+        // validate reamortization is only done on an active loan
+        if (!loan.getStatus().isActive()) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.supported.only.for.active.loans",
+                    "Loan reamortization can only be done on active loans", loan.getId());
+        }
+
+        // validate if there's already a re-amortization transaction for today
+        boolean isReAmortizationTransactionForTodayPresent = !loan
+                .getLoanTransactions(tx -> tx.getTypeOf().isReAmortize() && tx.getTransactionDate().equals(getBusinessLocalDate()))
+                .isEmpty();
+        if (isReAmortizationTransactionForTodayPresent) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.reamortize.transaction.already.present.for.today",
+                    "Loan reamortization can only be done once a day. There has already been a reamortization done for today",
+                    loan.getId());
+        }
     }
 
     public void validateUndoReAmortize(Loan loan, JsonCommand command) {
-        // TODO: implement
+        validateUndoReAmortizeBusinessRules(loan);
+    }
+
+    private void validateUndoReAmortizeBusinessRules(Loan loan) {
+        // validate if there's a reamortization transaction already
+        Optional<LoanTransaction> optionalReAmortizationTx = loan.getLoanTransactions(tx -> tx.getTypeOf().isReAmortize()).stream()
+                .min(Comparator.comparing(LoanTransaction::getTransactionDate));
+        if (optionalReAmortizationTx.isEmpty()) {
+            throw new GeneralPlatformDomainRuleException("error.msg.loan.reamortize.reamortization.transaction.missing",
+                    "Undoing a reamortization can only be done if there was a reamortization already", loan.getId());
+        }
+
+        // validate if there's no payment between the reamortization and today
+        LocalDate reAmortizationTxDate = optionalReAmortizationTx.get().getTransactionDate();
+        boolean repaymentExistsAfterReAmortization = !loan
+                .getLoanTransactions(
+                        tx -> tx.getTypeOf().isRepaymentType() && DateUtils.isAfter(reAmortizationTxDate, getBusinessLocalDate()))
```

Comment:
What is the repayment is on the same day, but it happens after the reamortization?

Reply:
Edge case, doesn't need to be handled for now.